crates/core/tedge_write/src/bin.rs

```diff
@@ -62,133 +56,34 @@ pub fn run(args: Args) -> anyhow::Result<()> {
-        .with_context(|| format!("Could not open temporary file `{temp_filepath}` for writing"))?;
-
-    // If the target file already exists, use the same permissions and uid/gid
-    if target_filepath.is_file() {
-        let target_metadata = target_filepath.metadata().with_context(|| {
-            format!("Could not fetch metadata of target file {target_filepath}")
-        })?;
-
-        let uid = target_metadata.uid();
-        let gid = target_metadata.gid();
-        chown(&temp_filepath, Some(uid), Some(gid))
-            .context("Could not set destination owner/group")?;
-
-        let target_permissions = target_metadata.permissions();
-        temp_file
-            .set_permissions(target_permissions)
-            .with_context(|| {
-                format!("Could not set permissions for temporary file {temp_filepath}")
-            })?;
-    }
-
-    let mut stdin = std::io::stdin().lock();
-    io::copy(&mut stdin, &mut temp_file)
-        .with_context(|| format!("Could not write to the temporary file `{temp_filepath}`"))?;
-
-    if let Err(e) = fs::rename(&temp_filepath, target_filepath)
-        .with_context(|| format!("Could not write to destination file `{target_filepath}`"))
-    {
-        let _ = fs::remove_file(&temp_filepath);
-        return Err(e);
-    }
-    Ok(())
-}
+    let uid = args
+        .user
+        .map(|u| uzers::get_user_by_name(&*u).with_context(|| format!("no such user: '{u}'")))
+        .transpose()?
+        .map(|u| u.uid());
 
-fn chown_by_user_and_group_name(
-    filepath: &Utf8Path,
-    user: Option<&str>,
-    group: Option<&str>,
-) -> anyhow::Result<()> {
-    // if user and group contain only digits, then they're ids
-    let user_id = user.and_then(|u| u.parse::<u32>().ok());
-    let group_id = group.and_then(|g| g.parse::<u32>().ok());
+    let gid = args
+        .group
+        .map(|g| uzers::get_group_by_name(&*g).with_context(|| format!("no such group: '{g}'")))
+        .transpose()?
+        .map(|g| g.gid());
 
-    let new_uid = match user {
-        Some(u) => {
-            if user_id.is_some() {
-                user_id
-            } else {
-                Some(
-                    uzers::get_user_by_name(u)
-                        .with_context(|| format!("User `{u}` does not exist"))?
-                        .uid(),
-                )
-            }
-        }
-        None => None,
-    };
+    // what permissions we want to set if the file doesn't exist
+    let permissions = MaybePermissions { uid, gid, mode };
 
-    let new_gid = match group {
-        Some(g) => {
-            if group_id.is_some() {
-                group_id
-            } else {
-                Some(
-                    uzers::get_group_by_name(g)
-                        .with_context(|| format!("Group `{g}` does not exist"))?
-                        .gid(),
-                )
-            }
-        }
-        None => None,
-    };
+    let src = std::io::stdin().lock();
```

Comment:
Not a remark on this PR, but a query on the initial design itself. Why was this plugin designed this way to read the input file content via stdin? If you could just share a link to some discussions on this point on the older issue/PR, that'd suffice.

Reply:
https://github.com/thin-edge/thin-edge.io/issues/2456

There was a requirement that tedge-write should not be able to read files that are not globally readable, e.g. private keys. This is naturally achieved when the calling side needs to open the file to pipe it into stdin.